lib/ssz_ex.ex

```diff
@@ -64,6 +77,95 @@ defmodule LambdaEthereumConsensus.SszEx do
     |> flatten_results_by(&Enum.join/1)
   end
 
+  defp decode_bitlist(bit_list, max_size) do
+    num_bytes = byte_size(bit_list)
+
+    with {:ok, len} <- get_bitlist_len(bit_list) do
+      cond do
+        div(len, 8) + 1 != num_bytes ->
+          {:error, "InvalidByteCount"}
+
+        len > max_size ->
+          {:error, "OutOfBounds"}
+
+        true ->
+          bytes_for_bit_len = max(1, div(len + 7, 8))
+          bit_size = bit_size(bit_list)
+          index = rem(len, 8)
+          skip = bit_size - index - 1
+          <<pre::bitstring-size(skip), _::size(1), rest::bitstring>> = bit_list
+
+          <<pre::binary-size(bytes_for_bit_len - 1), rest::bitstring>> =
+            <<pre::bitstring, 0::1, rest::bitstring>>
+
+          remain_bits_size = len - bit_size(pre)
+          <<_::size(8 - remain_bits_size), rest::bitstring>> = rest
+          formatted = <<pre::bitstring, rest::bitstring>>
+
+          if byte_size(formatted) > bytes_for_bit_len do
+            <<truncated::binary-size(bytes_for_bit_len), _rest::bitstring>> = formatted
+            {:ok, truncated}
+          else
+            {:ok, formatted}
+          end
+      end
+    end
+  end
+
+  defp get_bitlist_len(bit_list) do
+    :binary.bin_to_list(bit_list)
+    |> Enum.with_index()
+    |> Enum.reverse()
+    |> Enum.find_value({:error, "missing_length_information"}, fn {byte, byte_index} ->
+      if for(<<(bit::1 <- <<byte>>)>>, do: bit) |> Enum.any?(&(&1 == 1)) do
+        {:ok, {byte, byte_index}}
+      else
+        {:error, "missing_length_information"}
+      end
+    end)
+    |> case do
+      {:ok, {byte, byte_index}} ->
+        leading_zeros =
+          for(<<(bit::1 <- <<byte>>)>>, do: bit)
+          |> Enum.take_while(&(&1 == 0))
+          |> Enum.count()
+
+        {:ok, byte_index * 8 + 7 - leading_zeros}
+
+      error ->
+        error
+    end
+  end
+
+  defp decode_bitvector(bit_vector, size) when bit_size(bit_vector) == size,
+    do: {:ok, BitVector.new(bit_vector, size)}
+
+  defp decode_bitvector(_bit_vector, _size), do: {:error, "invalid bit_vector length"}
+
+  defp encode_bitlist(bit_list, max_size) do
+    len = bit_size(bit_list)
+    bytes_len = byte_size(bit_list)
+    bytes_for_bit_len = max(1, div(len + 1 + 7, 8))
+
+    if len > max_size do
+      {:error, "ExcessBits"}
+    else
+      if bytes_for_bit_len > bytes_len do
+        append_size = (bytes_for_bit_len - bytes_len) * @bits_per_byte
+        resized = <<bit_list::bitstring, 0::size(append_size)>>
+        skip = bit_size(resized) - 1
+        <<pre::bitstring-size(skip), _::size(1), rest::bitstring>> = resized
+        {:ok, <<pre::bitstring, 1::1, rest::bitstring>>}
+      else
+        <<pre::binary-size(bytes_len - 1), last::bitstring>> = bit_list
+        {:ok, <<pre::bitstring, 0::size(8 - bit_size(last) - 1), last::bitstring, 1::1>>}
+      end
```

Comment:
Can't we replace this whole logic with something simpler? If I understand correctly, we only need to append a 1 and many 0s. Can't we just do something like this?

```elixir
padding = rem(bit_size(bit_list), 8)
<<bit_list-bitstring, 1::1, 0::size(padding)>>
```

Reply:
It depends. For example, for `<<15::size(4)>>` it would be only necessary to add  `1` but for `<<255>>` it would be necessary to add `0::size(7), 1::1` :thinking:

Reply:
Yes, my original implementation misinterpreted how the little-endian-bit-order worked. I added a new suggestion after doing some research.